File: backend/src/storage.ts
import fs from 'fs'
import path from 'path'
import crypto from 'crypto'
import { configManager } from './config.js'
import { DownloadOptions } from './builder.js'

export interface Preset {
  id: string
  name: string
  description: string
  icon: string
  isBuiltin: boolean
  options: DownloadOptions
  createdAt: string
}

export interface StoredJob {
  id: string
  url: string
  title: string
  thumbnail: string
  duration: number
  durationString: string
  uploader: string
  status: string
  outputFile: string
  files: string[]
  downloadedBytes: number
  totalBytes: number
  options: DownloadOptions
  createdAt: string
  completedAt?: string
  errorMessage?: string
}

class StorageManager {
  private historyPath: string
  private presetsPath: string
  private jobs: Map<string, StoredJob> = new Map()
  private presets: Map<string, Preset> = new Map()

  constructor() {
    const dataDir = configManager.getDataDir()
    this.historyPath = path.join(dataDir, 'history.json')
    this.presetsPath = path.join(dataDir, 'presets.json')
    this.load()
  }

  private load() {
    // History - only keep jobs whose files actually exist on disk
    try {
      if (fs.existsSync(this.historyPath)) {
        const raw = fs.readFileSync(this.historyPath, 'utf-8')
        const list: StoredJob[] = JSON.parse(raw)
        for (const j of list) {
          const fileExists =
            (j.outputFile && fs.existsSync(j.outputFile)) ||
            (Array.isArray(j.files) && j.files.some((f) => f && fs.existsSync(f)))

          if (!fileExists) {
            continue
          }

          this.jobs.set(j.id, j)
        }
        this.saveHistory()
      }
    } catch {
      // Ignored
    }

    // Presets
    try {
      if (fs.existsSync(this.presetsPath)) {
        const raw = fs.readFileSync(this.presetsPath, 'utf-8')
        const list: Preset[] = JSON.parse(raw)
        for (const p of list) {
          this.presets.set(p.id, p)
        }
      }
    } catch {
      // Ignored
    }

    if (this.presets.size === 0) {
      this.seedDefaultPresets()
    }
  }

  private seedDefaultPresets() {
    const defaults: Preset[] = [
      {
        id: 'preset-best-video',
        name: 'Best Quality Video (4K/HD)',
        description: 'Downloads the highest available video & audio, remuxed to MP4 with embedded metadata and thumbnail',
        icon: 'film',
        isBuiltin: true,
        options: {
          url: '',
          mode: 'video',
          videoQuality: 'best',
          videoContainer: 'mp4',
          embedMetadata: true,
          embedThumbnail: true,
          embedChapters: true,
          sponsorBlockAction: 'remove',
        },
        createdAt: new Date().toISOString(),
      },
      {
        id: 'preset-1080p-mp4',
        name: 'Standard 1080p MP4',
        description: 'Maximum 1080p Full HD MP4 for broad compatibility and fast download speeds',
        icon: 'monitor',
        isBuiltin: true,
        options: {
          url: '',
          mode: 'video',
          videoQuality: '1080',
          videoContainer: 'mp4',
          embedMetadata: true,
          embedThumbnail: true,
        },
        createdAt: new Date().toISOString(),
      },
      {
        id: 'preset-mp3-hq',
        name: 'High Quality MP3 (320 kbps)',
        description: 'Extracts audio converted to 320 kbps MP3 with album art and ID3 metadata tags',
        icon: 'music',
        isBuiltin: true,
        options: {
          url: '',
          mode: 'audio',
          audioFormat: 'mp3',
          audioQuality: '320k',
          embedMetadata: true,
          embedThumbnail: true,
        },
        createdAt: new Date().toISOString(),
      },
      {
        id: 'preset-flac-lossless',
        name: 'Lossless Audio (FLAC)',
        description: 'Extracts pristine lossless FLAC audio with high resolution embedded metadata',
        icon: 'disc',
        isBuiltin: true,
        options: {
          url: '',
          mode: 'audio',
          audioFormat: 'flac',
          audioQuality: '0',
          embedMetadata: true,
          embedThumbnail: true,
        },
        createdAt: new Date().toISOString(),
      },
      {
        id: 'preset-archival-mkv',
        name: 'Full Archival MKV',
        description: 'Preserves all subtitles, auto-subs, original streams, chapters, and metadata in an MKV container',
        icon: 'archive',
        isBuiltin: true,
        options: {
          url: '',
          mode: 'video',
          videoQuality: 'best',
          videoContainer: 'mkv',
          embedMetadata: true,
          embedThumbnail: true,
          embedChapters: true,
          embedSubtitles: true,
          autoSubtitles: true,
          subtitleLanguages: 'all',
          sponsorBlockAction: 'mark',
        },
        createdAt: new Date().toISOString(),
      },
    ]

    for (const d of defaults) {
      this.presets.set(d.id, d)
    }
    this.savePresets()
  }

  public saveJob(job: StoredJob) {
    this.jobs.set(job.id, job)
    this.saveHistory()
  }

  public deleteJob(id: string) {
    this.jobs.delete(id)
    this.saveHistory()
  }

  public renameJobFile(oldPath: string, newPath: string) {
    const oldNorm = path.resolve(oldPath).toLowerCase()
    const newName = path.basename(newPath)
    for (const job of this.jobs.values()) {
      let matched = false
      if (job.outputFile && path.resolve(job.outputFile).toLowerCase() === oldNorm) {
        job.outputFile = newPath
        matched = true
      }
      if (Array.isArray(job.files)) {
        job.files = job.files.map((f) => {
          if (f && path.resolve(f).toLowerCase() === oldNorm) {
            matched = true
            return newPath
          }
          return f
        })
      }
      if (matched) {
        job.title = newName
      }
    }
    this.saveHistory()
  }

  public clearCompleted() {
    for (const [id, job] of this.jobs.entries()) {
      if (job.status === 'completed' || job.status === 'failed' || job.status === 'cancelled') {
        this.jobs.delete(id)
      }
    }
    this.saveHistory()
  }

  public getAllJobs(): StoredJob[] {
    return Array.from(this.jobs.values())
  }

  public getPresets(): Preset[] {
    return Array.from(this.presets.values())
  }

  public savePreset(preset: Partial<Preset>): Preset {
    const id = preset.id || `preset-${crypto.randomBytes(4).toString('hex')}`
    const full: Preset = {
      id,
      name: preset.name || 'Custom Preset',
      description: preset.description || '',
      icon: preset.icon || 'sparkles',
      isBuiltin: Boolean(preset.isBuiltin),
      options: preset.options || { url: '', mode: 'video' },
      createdAt: preset.createdAt || new Date().toISOString(),
    }
    this.presets.set(id, full)
    this.savePresets()
    return full
  }

  public deletePreset(id: string): boolean {
    const p = this.presets.get(id)
    if (!p || p.isBuiltin) return false
    this.presets.delete(id)
    this.savePresets()
    return true
  }

  private saveHistory() {
    try {
      const list = Array.from(this.jobs.values())
      fs.writeFileSync(this.historyPath, JSON.stringify(list, null, 2), 'utf-8')
    } catch {
      // Ignored
    }
  }

  private savePresets() {
    try {
      const list = Array.from(this.presets.values())
      fs.writeFileSync(this.presetsPath, JSON.stringify(list, null, 2), 'utf-8')
    } catch {
      // Ignored
    }
  }
}

export const storageManager = new StorageManager()

File: backend/pkg/storage/storage.go
package storage

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sync"
	"time"

	"github.com/google/uuid"
	"ytdlpgui/backend/pkg/config"
	"ytdlpgui/backend/pkg/ytdlp"
)

type Preset struct {
	ID          string                `json:"id"`
	Name        string                `json:"name"`
	Description string                `json:"description"`
	Icon        string                `json:"icon"`
	IsBuiltin   bool                  `json:"isBuiltin"`
	Options     ytdlp.DownloadOptions `json:"options"`
	CreatedAt   time.Time             `json:"createdAt"`
}

type StoredJob struct {
	ID              string                `json:"id"`
	URL             string                `json:"url"`
	Title           string                `json:"title"`
	Thumbnail       string                `json:"thumbnail"`
	Duration        int64                 `json:"duration"`
	DurationString  string                `json:"durationString"`
	Uploader        string                `json:"uploader"`
	Status          string                `json:"status"`
	OutputFile      string                `json:"outputFile"`
	Files           []string              `json:"files"`
	DownloadedBytes int64                 `json:"downloadedBytes"`
	TotalBytes      int64                 `json:"totalBytes"`
	Options         ytdlp.DownloadOptions `json:"options"`
	CreatedAt       time.Time             `json:"createdAt"`
	CompletedAt     *time.Time            `json:"completedAt,omitempty"`
	ErrorMessage    string                `json:"errorMessage,omitempty"`
}

type Store struct {
	mu           sync.RWMutex
	historyPath  string
	presetsPath  string
	jobs         map[string]StoredJob
	presets      map[string]Preset
}

var (
	storeInstance *Store
	storeOnce     sync.Once
)

func GetStore() *Store {
	storeOnce.Do(func() {
		cfgMgr := config.GetManager()
		dataDir := cfgMgr.GetDataDir()

		storeInstance = &Store{
			historyPath: filepath.Join(dataDir, "history.json"),
			presetsPath: filepath.Join(dataDir, "presets.json"),
			jobs:        make(map[string]StoredJob),
			presets:     make(map[string]Preset),
		}
		storeInstance.load()
	})
	return storeInstance
}

func (s *Store) load() {
	s.mu.Lock()
	defer s.mu.Unlock()

	// Load History
	if data, err := os.ReadFile(s.historyPath); err == nil {
		var list []StoredJob
		if err := json.Unmarshal(data, &list); err == nil {
			for _, item := range list {
				s.jobs[item.ID] = item
			}
		}
	}

	// Load Presets
	if data, err := os.ReadFile(s.presetsPath); err == nil {
		var list []Preset
		if err := json.Unmarshal(data, &list); err == nil {
			for _, item := range list {
				s.presets[item.ID] = item
			}
		}
	}

	// Seed built-in presets if empty
	if len(s.presets) == 0 {
		s.seedDefaultPresets()
	}
}

func (s *Store) seedDefaultPresets() {
	defaults := []Preset{
		{
			ID:          "preset-best-video",
			Name:        "Best Quality Video (4K/HD)",
			Description: "Downloads the highest available video & audio, remuxed to MP4 with embedded metadata and thumbnail",
			Icon:        "film",
			IsBuiltin:   true,
			Options: ytdlp.DownloadOptions{
				Mode:               "video",
				VideoQuality:       "best",
				VideoContainer:     "mp4",
				EmbedMetadata:      true,
				EmbedThumbnail:     true,
				EmbedChapters:      true,
				SponsorBlockAction: "remove",
			},
			CreatedAt: time.Now(),
		},
		{
			ID:          "preset-1080p-mp4",
			Name:        "Standard 1080p MP4",
			Description: "Maximum 1080p Full HD MP4 for broad compatibility and fast download speeds",
			Icon:        "monitor",
			IsBuiltin:   true,
			Options: ytdlp.DownloadOptions{
				Mode:           "video",
				VideoQuality:   "1080",
				VideoContainer: "mp4",
				EmbedMetadata:  true,
				EmbedThumbnail: true,
			},
			CreatedAt: time.Now(),
		},
		{
			ID:          "preset-mp3-hq",
			Name:        "High Quality MP3 (320 kbps)",
			Description: "Extracts audio converted to 320 kbps MP3 with album art and ID3 metadata tags",
			Icon:        "music",
			IsBuiltin:   true,
			Options: ytdlp.DownloadOptions{
				Mode:           "audio",
				AudioFormat:    "mp3",
				AudioQuality:   "320k",
				EmbedMetadata:  true,
				EmbedThumbnail: true,
			},
			CreatedAt: time.Now(),
		},
		{
			ID:          "preset-flac-lossless",
			Name:        "Lossless Audio (FLAC)",
			Description: "Extracts pristine lossless FLAC audio with high resolution embedded metadata",
			Icon:        "disc",
			IsBuiltin:   true,
			Options: ytdlp.DownloadOptions{
				Mode:           "audio",
				AudioFormat:    "flac",
				AudioQuality:   "0",
				EmbedMetadata:  true,
				EmbedThumbnail: true,
			},
			CreatedAt: time.Now(),
		},
		{
			ID:          "preset-archival-mkv",
			Name:        "Full Archival MKV",
			Description: "Preserves all subtitles, auto-subs, original streams, chapters, and metadata in an MKV container",
			Icon:        "archive",
			IsBuiltin:   true,
			Options: ytdlp.DownloadOptions{
				Mode:               "video",
				VideoQuality:       "best",
				VideoContainer:     "mkv",
				EmbedMetadata:      true,
				EmbedThumbnail:     true,
				EmbedChapters:      true,
				EmbedSubtitles:     true,
				AutoSubtitles:      true,
				SubtitleLanguages:  "all",
				SponsorBlockAction: "mark",
			},
			CreatedAt: time.Now(),
		},
	}

	for _, p := range defaults {
		s.presets[p.ID] = p
	}
	_ = s.savePresetsUnsafe()
}

func (s *Store) SaveJob(job StoredJob) {
	s.mu.Lock()
	defer s.mu.Unlock()

	s.jobs[job.ID] = job
	_ = s.saveHistoryUnsafe()
}

func (s *Store) DeleteJob(id string) {
	s.mu.Lock()
	defer s.mu.Unlock()

	delete(s.jobs, id)
	_ = s.saveHistoryUnsafe()
}

func (s *Store) ClearCompleted() {
	s.mu.Lock()
	defer s.mu.Unlock()

	for id, job := range s.jobs {
		if job.Status == "completed" || job.Status == "failed" || job.Status == "cancelled" {
			delete(s.jobs, id)
		}
	}
	_ = s.saveHistoryUnsafe()
}

func (s *Store) GetAllJobs() []StoredJob {
	s.mu.RLock()
	defer s.mu.RUnlock()

	list := make([]StoredJob, 0, len(s.jobs))
	for _, j := range s.jobs {
		list = append(list, j)
	}
	return list
}

func (s *Store) GetPresets() []Preset {
	s.mu.RLock()
	defer s.mu.RUnlock()

	list := make([]Preset, 0, len(s.presets))
	for _, p := range s.presets {
		list = append(list, p)
	}
	return list
}

func (s *Store) SavePreset(preset Preset) Preset {
	s.mu.Lock()
	defer s.mu.Unlock()

	if preset.ID == "" {
		preset.ID = "preset-" + uuid.New().String()[:8]
		preset.CreatedAt = time.Now()
	}
	s.presets[preset.ID] = preset
	_ = s.savePresetsUnsafe()
	return preset
}

func (s *Store) DeletePreset(id string) bool {
	s.mu.Lock()
	defer s.mu.Unlock()

	if p, ok := s.presets[id]; ok {
		if p.IsBuiltin {
			return false // Do not delete builtin presets
		}
		delete(s.presets, id)
		_ = s.savePresetsUnsafe()
		return true
	}
	return false
}

func (s *Store) saveHistoryUnsafe() error {
	list := make([]StoredJob, 0, len(s.jobs))
	for _, j := range s.jobs {
		list = append(list, j)
	}
	data, err := json.MarshalIndent(list, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(s.historyPath, data, 0644)
}

func (s *Store) savePresetsUnsafe() error {
	list := make([]Preset, 0, len(s.presets))
	for _, p := range s.presets {
		list = append(list, p)
	}
	data, err := json.MarshalIndent(list, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(s.presetsPath, data, 0644)
}
